### ISSUE TITLE
Some files can get corrupted when using multiple OsmAnd versions in parallel

### ISSUE THREAD

**Issue by scaidermern** [osmandapp/Osmand#3012]
When using multiple OsmAnd versions in parallel (e.g. stable and nightly) and using the same data folder (for reducing required storage) some files can get corrupted. This seems especially apply to files which get often re-written by OsmAnd, for example _regions.ocbf_, _poi_types.xml_ and style files located in the _rendering/_ directory like _default.renderer.xml_.

With corrupted files several different exceptions might occur on startup.

This exception seems to belong to a corrupted _regions.ocbf_:
```
08-27 18:33:50.358 E/net.osmand(15330): AppInitializer While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either than the input has been truncated or that an embedded message misreported its own length.
08-27 18:33:50.358 E/net.osmand(15330): com.google.protobuf.InvalidProtocolBufferException: While parsing a protocol message, the input ended unexpectedly in the middle of a field.  This could mean either than the input has been truncated or that an embedded message misreported its own length.
08-27 18:33:50.358 E/net.osmand(15330): 	at com.google.protobuf.CodedInputStream.refillBuffer(CodedInputStream.java:745)
08-27 18:33:50.358 E/net.osmand(15330): 	at com.google.protobuf.CodedInputStream.readRawByte(CodedInputStream.java:798)
08-27 18:33:50.358 E/net.osmand(15330): 	at com.google.protobuf.CodedInputStream.readRawVarint32(CodedInputStream.java:384)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.binary.BinaryMapIndexReader.readMapDataObject(BinaryMapIndexReader.java:1159)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.binary.BinaryMapIndexReader.readMapDataBlocks(BinaryMapIndexReader.java:926)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.binary.BinaryMapIndexReader.searchMapIndex(BinaryMapIndexReader.java:878)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.binary.BinaryMapIndexReader.searchMapIndex(BinaryMapIndexReader.java:822)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.map.OsmandRegions.iterateOverAllObjects(OsmandRegions.java:516)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.map.OsmandRegions.prepareFile(OsmandRegions.java:127)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.plus.AppInitializer.indexRegionsBoundaries(AppInitializer.java:296)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.plus.AppInitializer.startApplicationBackground(AppInitializer.java:502)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.plus.AppInitializer.access$200(AppInitializer.java:74)
08-27 18:33:50.358 E/net.osmand(15330): 	at net.osmand.plus.AppInitializer$9.run(AppInitializer.java:738)
08-27 18:33:50.358 E/net.osmand(15330): 	at java.lang.Thread.run(Thread.java:818)
```

This exception occurs if the currently selected rendering style sheet (.e.g. _default.renderer.xml_) got broken:
```
08-27 18:53:05.380 E/net.osmand(22767): RendererRegistry Error loading renderer
08-27 18:53:05.380 E/net.osmand(22767): org.xmlpull.v1.XmlPullParserException: name expected (position:START_TAG <case>@3324:11 in java.io.InputStreamReader@33a67b63) 
08-27 18:53:05.380 E/net.osmand(22767): 	at org.kxml2.io.KXmlParser.checkRelaxed(KXmlParser.java:305)
08-27 18:53:05.380 E/net.osmand(22767): 	at org.kxml2.io.KXmlParser.readName(KXmlParser.java:1547)
08-27 18:53:05.380 E/net.osmand(22767): 	at org.kxml2.io.KXmlParser.parseStartTag(KXmlParser.java:1090)
08-27 18:53:05.380 E/net.osmand(22767): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:372)
08-27 18:53:05.380 E/net.osmand(22767): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
08-27 18:53:05.380 E/net.osmand(22767): 	at net.osmand.plus.render.RendererRegistry.loadRenderer(RendererRegistry.java:111)
08-27 18:53:05.380 E/net.osmand(22767): 	at net.osmand.plus.render.RendererRegistry.getRenderer(RendererRegistry.java:79)
08-27 18:53:05.380 E/net.osmand(22767): 	at net.osmand.plus.dialogs.ConfigureMapMenu$2$1.onClick(ConfigureMapMenu.java:316)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.support.v7.app.AlertController$AlertParams$3.onItemClick(AlertController.java:959)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.widget.AdapterView.performItemClick(AdapterView.java:305)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.widget.AbsListView.performItemClick(AbsListView.java:1148)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.widget.AbsListView$PerformClick.run(AbsListView.java:3059)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.widget.AbsListView$3.run(AbsListView.java:3866)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.os.Handler.handleCallback(Handler.java:739)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.os.Handler.dispatchMessage(Handler.java:95)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.os.Looper.loop(Looper.java:135)
08-27 18:53:05.380 E/net.osmand(22767): 	at android.app.ActivityThread.main(ActivityThread.java:5294)
08-27 18:53:05.380 E/net.osmand(22767): 	at java.lang.reflect.Method.invoke(Native Method)
08-27 18:53:05.380 E/net.osmand(22767): 	at java.lang.reflect.Method.invoke(Method.java:372)
08-27 18:53:05.380 E/net.osmand(22767): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
08-27 18:53:05.380 E/net.osmand(22767): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
```

And another exception if _poi_types.xml_ gets corrupted:
```
08-20 08:09:06.635 E/net.osmand(9285): MapRenderingTypes Unexpected error
08-20 08:09:06.635 E/net.osmand(9285): org.xmlpull.v1.XmlPullParserException: Attr.value missing f. vah (position:START_TAG <poi_additional name='indoor_yes' tag='indoor' vah='no'>@453:52 in java.io.InputStreamReader@231afd13) 
08-20 08:09:06.635 E/net.osmand(9285): 	at org.kxml2.io.KXmlParser.checkRelaxed(KXmlParser.java:305)
08-20 08:09:06.635 E/net.osmand(9285): 	at org.kxml2.io.KXmlParser.parseStartTag(KXmlParser.java:1131)
08-20 08:09:06.635 E/net.osmand(9285): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:372)
08-20 08:09:06.635 E/net.osmand(9285): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
08-20 08:09:06.635 E/net.osmand(9285): 	at net.osmand.osm.MapPoiTypes.init(MapPoiTypes.java:279)
08-20 08:09:06.635 E/net.osmand(9285): 	at net.osmand.plus.AppInitializer.initPoiTypes(AppInitializer.java:308)
08-20 08:09:06.635 E/net.osmand(9285): 	at net.osmand.plus.AppInitializer.startApplicationBackground(AppInitializer.java:489)
08-20 08:09:06.635 E/net.osmand(9285): 	at net.osmand.plus.AppInitializer.access$200(AppInitializer.java:72)
08-20 08:09:06.635 E/net.osmand(9285): 	at net.osmand.plus.AppInitializer$9.run(AppInitializer.java:738)
08-20 08:09:06.635 E/net.osmand(9285): 	at java.lang.Thread.run(Thread.java:818)
08-20 08:09:06.635 W/System.err(9285): 	at net.osmand.osm.MapPoiTypes.init(MapPoiTypes.java:279)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.initPoiTypes(AppInitializer.java:308)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.startApplicationBackground(AppInitializer.java:489)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.access$200(AppInitializer.java:72)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer$9.run(AppInitializer.java:738)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.osm.MapPoiTypes.init(MapPoiTypes.java:354)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.initPoiTypes(AppInitializer.java:308)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.startApplicationBackground(AppInitializer.java:489)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer.access$200(AppInitializer.java:72)
08-20 08:09:06.636 W/System.err(9285): 	at net.osmand.plus.AppInitializer$9.run(AppInitializer.java:738)
08-20 08:09:06.637 W/System.err(9285): 	at net.osmand.osm.MapPoiTypes.init(MapPoiTypes.java:279)
```

The error messages that OsmAnd will display to the user aren't particularly helpful. Yet fixing these issues is for the user very easy: just delete the corrupted file (_regions.ocbf_, _poi_types.xml_, _default.renderer.xml_ etc.). They will get re-created by OsmAnd on startup.

Since the manual fix is very easy it seems to be a good idea to implement an automated fix in OsmAnd. If loading of one of these files fails then just re-create it and try loading it again (once). This should make OsmAnd more robust against such errors.

**Comment by vshcherb** [comment]
Looks quite bad, though I'm also using multiple versions in parallel could you please describe what does it mean in parallel? Do you start the application in the same time? 

The files you mentionned indeed are rewritten each new *installation* , though they should not be corrupted.

I'm afraid we need to understand how do we get corrupted files, cause introducing auto-recovery could be quite difficult algorithm.

**Comment by scaidermern** [comment]
> could you please describe what does it mean in parallel? Do you start the application in the same time?

Not exactly. But it may happen for example that if the nightly has some bug I'm pressing the home button and start the stable instead. Or vice versa. In this case the previous OsmAnd can be still running and closed at a later point in time by Android.

> The files you mentioned indeed are rewritten each new installation , though they should not be corrupted.
Oh, just on installation? Is that really *while installing*? Because at this point in time OsmAnd doesn't know the data directory, does it? Or is it at the first start of the newly installed version?

So far I always imagined these files get corrupted during exiting of OsmAnd when one app writes to the files while the other one has it still opened or something like this. However on a typical Linux system this doesn't lead to file corruption since the file descriptors should differ.

**Comment by vshcherb** [comment]
That's true, though sdcard could run different FS (FAT32 for example). Though it still strange cause default.render.xml is not kept opened. So as I see it is vice versa map files are not corrupted but they are opened, other files are read at the start and they are corrupted. So it is strange

**Comment by scaidermern** [comment]
I played a little around and could reproduce the _poi_types.xml_ exception.

What I did: Opened OsmAnd+ (stable). Pressed home button. Downloaded and installed latest nightly. Opened nightly. Pressed home button. Opened OsmAnd+ (stable) again. The exception occurred right on startup.

Now _poi_types.xml_ contains garbage somewhere in between:

![photo_2016-08-28_16-57-54](https://cloud.githubusercontent.com/assets/1040784/18034569/9dc02460-6d40-11e6-965e-37ba342abc18.jpg)

And at a later position the garbage vanishes and the XML content appears again (which is really strange in my opinion):

![photo_2016-08-28_16-57-57](https://cloud.githubusercontent.com/assets/1040784/18034573/b813522e-6d40-11e6-9742-cea133450eca.jpg)

Here is the [Log](https://github.com/osmandapp/Osmand/files/441185/2016-08-28-16-49-02.txt).

I know this might not be a typical work flow but I already encountered these three exceptions multiple times so they definitely happen when playing around with different OsmAnd versions.

**Comment by scaidermern** [comment]
After manually deleting _poi_types.xml_ I noticed that _default.renderer.xml_ also got corrupted:

```
08-28 17:04:38.196 E/net.osmand(29846): RendererRegistry Error loading renderer
08-28 17:04:38.196 E/net.osmand(29846): org.xmlpull.v1.XmlPullParserException: unterminated entity ref (position:START_TAG <renderingConstant name='otherAmenitiesTextHaloColorDay'>@1430:104 in java.io.InputStreamReader@1b722e50) 
08-28 17:04:38.196 E/net.osmand(29846): 	at org.kxml2.io.KXmlParser.readEntity(KXmlParser.java:1219)
08-28 17:04:38.196 E/net.osmand(29846): 	at org.kxml2.io.KXmlParser.readValue(KXmlParser.java:1401)
08-28 17:04:38.196 E/net.osmand(29846): 	at org.kxml2.io.KXmlParser.parseStartTag(KXmlParser.java:1122)
08-28 17:04:38.196 E/net.osmand(29846): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:372)
08-28 17:04:38.196 E/net.osmand(29846): 	at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.render.RendererRegistry.loadRenderer(RendererRegistry.java:111)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.render.RendererRegistry.getRenderer(RendererRegistry.java:79)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.render.RendererRegistry.initRenderers(RendererRegistry.java:226)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.AppInitializer.startApplicationBackground(AppInitializer.java:493)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.AppInitializer.access$200(AppInitializer.java:72)
08-28 17:04:38.196 E/net.osmand(29846): 	at net.osmand.plus.AppInitializer$9.run(AppInitializer.java:738)
08-28 17:04:38.196 E/net.osmand(29846): 	at java.lang.Thread.run(Thread.java:818)
```

This time the garbage is up to the end of the file:

![photo_2016-08-28_17-08-46](https://cloud.githubusercontent.com/assets/1040784/18034647/4c232ed4-6d42-11e6-836d-b919e76059a3.jpg)

![photo_2016-08-28_17-08-43](https://cloud.githubusercontent.com/assets/1040784/18034646/47cfd8e6-6d42-11e6-88cc-8c75391ee216.jpg)

Here is the [log](https://github.com/osmandapp/Osmand/files/441192/2016-08-28-17-05-07.txt) (it also includes the previous log).

**Comment by vshcherb** [comment]
This not a very typical workflow but still typical for developers. Tried to reproduce on my device no luck yet.

**Comment by stephan75** [comment]
Maybe the sd card has a hardware failure?

Can you reproduce with another card, or when using internal memory?

**Comment by scaidermern** [comment]
I tried to reproduce the problem with internal memory (by configuring the same path for both OsmAnd apps) but couldn't reproduce it. I tried a different sdcard but couldn't reproduce it. I run fsck on the original sdcard, no error was found. I tried the original sdcard again but couldn't reproduce it either.

This is strange since I already encountered these problems multiple times. But only with OsmAnd, I never had problems with other files on the external sd card. Maybe it requires proper timing for reproducing the error. I will report back if the problem appears again.
